MySQL APB example

codificat opened the pull request:
An example APB to deploy MySQL using the rhscl images.

Still WIP but submitting for feedback.

Known pending tasks:

- [x] confirm the *plan* approach to persistent storage. Revert to bool param if not yet ready
- [x] consider using a template for the DC instead of 2 separate dev/prod DCs
- [x] adjust `/etc/apb/bind-creds` so that it works under a restricted SCC
- [x] update `com.redhat.apb.spec` with the latest changes
- [ ] squash and rebase

codificat committed:
Add MySQL example APB

An Ansible Playbook Bundle to deploy MySQL using the SCL images

codificat committed:
Use OPENSHIFT_TOKEN instead of user and password

This matches the current apb-base login setup

codificat committed:
Populate defaults for all parameters

Required for the standalone docker run use case (i.e. out of ASB
invocation)

codificat committed:
Fix config change trigger typo

codificat committed:
Use fixed app name

codificat committed:
Environment variables to set params, extra-vars docs

codificat committed:
Fix image name to include rhscl

codificat committed:
Add relevant defaults to the deprovision role

codificat committed:
Also include rhscl in the name of the APB

codificat committed:
Updated Dockerfile with current apb

codificat committed:
Add health checks to the deployed container

codificat committed:
Fix variant name in travis

codificat committed:
Fix apb name in a comment

Basically an excuse to trigger a travis test

codificat committed:
Move ephemeral vs persistent storage to a plan option

Offer different plans, which determine if persistent or ephemeral storage will
be provided

codificat committed:
Add default plan to roles' defaults

Make sure there's a default plan set for the standalone invocation case.

codificat committed:
Improve mysql tests

Make the openshift-test in the MySQL APB more specific to this APB so it tests
provision and deprovision of the two available plans (ephemeral and PVC)

codificat committed:
Remove backup files

codificat committed:
Use a state var, unify [de]provision in one role

codificat committed:
Use the new way to pass credentials to the broker

codificat committed:
Minor README improvements

codificat committed:
Only process credentials during provisioning

codificat committed:
Remove leftover state from pre-state var usage

Left hardcoded state when introducing the 'state' var and reusing roles for both
provision and deprovision

jcpowermac commented:
LGTM

tchughesiv commented on rhscl-mysql-apb/Makefile:
```diff
@@ -0,0 +1,89 @@
+#	oc describe dc/${IMAGE_NAME}
+#	oc logs -f dc/${IMAGE_NAME}
+# FIXME: it's creating a pod, not a Job. Also fix the 'delete pod's below
+# FIXME: wait for the Job to complete with something better than a sleep/logs
```
+1 ... I'm going to research this a bit.

tchughesiv replied:
@codificat here's what i just did for hello-world-apb... moved to pod run and added logic to watch pod for completion...
https://github.com/fusor/apb-examples/commit/d84b7666cece64f481b17d6178203809697b7f21

codificat replied:
Thanks @tchughesiv - I'm pushing now an alternative approach using `--attach=true` in the `oc run` invocation. This should take care of showing the logs of the APB, waiting for its termination and exit with a matching exit code

codificat committed:
Remove the <BIND_CREDENTIALS> tag

codificat committed:
Updated tests for MySQL APB

Use a pod instead of a job (via --restart=Never in oc run).
Also sleep a bit after 'oc delete' to let it clean up, to avoid:

oc run rhscl-mysql-apb [...] -- provision -e namespace=mysql -e plan=prod
Error from server (AlreadyExists): object is being deleted: pods "rhscl-mysql-apb" already exists

codificat committed:
Address test FIXMEs by using attach on oc run

dymurray commented:
@codificat I merged in two other APBs that probably conflicted with your Makefile and Travis changes. Please rebase and I will start testing.

codificat committed:
Update creds file for mysql apb to /var/tmp

codificat commented:
Thanks @dymurray - Rebased and updated. Travis tests pass (which shows that they are not complete because they don't report a mismatch between the current apb-base and this image in bind-creds location...)

The main question I have is: this is using a plan to determine storage reqs. Are you ok with that or should we wait a bit more? (the commit history contains the previous implementation using a bool param instead of a plan, can revert to that)

I'll squash before merging

dymurray commented:
After talking with @codificat and @eriknelson our plan is to leave this PR OPEN until we implement plans in the Broker. @eriknelson will use this APB to test out the broker's implementation and this PR will be updated to match the final design.

eriknelson commented:
I'm marking this as a WIP since the broker does not support plans yet and this APB may require some minor changes as we build things out on the broker side. I'm hoping to have plan support in the broker sometime in the next week so we'll have a broker that can actually deploy this.

Thank you @codificat, super helpful.

codificat committed:
Updates as per the latest ASB improvements

Changes include:
 - shared plan params in apb.yml
 - remove non-needed from defaults (namespace, plan)
 - use apb.asb-modules

codificat committed:
Update testing to pass the plan as _apb_plan_id

codificat committed:
Updated Dockerfile with encoded current apb.yml

codificat committed:
Update README to match recent changes

codificat commented:
@dymurray I believe this is getting closer to be able to merge. I updated it to match recent changes in plan mgmt.

Known pending things:

- travis tests are currently failing (since credentials passing switched to exec-based) because there's no broker that gets the bind credentials from the APB. I think this would fail with any other bindable APB that had similar tests in place... I'll probably update Makefiles on a separate PR if that's ok?

- Rebase and squash all these commits before merging (I've been keeping history here to be able to look at progress/look back)

At this point though I'd like to get your thoughts about the content of the APB itself.

Thanks!

dymurray commented:
@codificat sorry for the delay. I tested your APB and it worked great on my end. I was able to deploy both plans and verified mysql came up. Please squash and rebase and I am happy to merge. I am fine with updating the makefile after.
